xdsl/printer.py

```diff
@@ -252,20 +250,13 @@ def print_region(self, region: Region) -> None:
         if not isinstance(region, Region):
             raise TypeError('Expected a Region; got %s' %
                             type(region).__name__)
-        if len(region.blocks) == 0:
-            self.print("{}")
-            return
 
-        if len(region.blocks) == 1 and len(region.blocks[0].args) == 0:
-            self.print("{")
-            self._print_ops(region.blocks[0].ops)
-            self.print("}")
-            return
+        print_block_name = len(region.blocks) != 1
```

Comment:
So this is something we need to be clear about.
The default generic format *always* prints the block name.
Whereas the custom format regions do not print the first block if it has no arguments.
Let's try to directly correspond to the syntax here, so we won't have to change these things twice? Or we should always print it, and then we'll see how we'll change it for the custom one.

Reply:
> The default generic format always prints the block name.

That seems to be a clear statement, no?

Reply:
Hmm, I've never seen a region with a single block that printed the block name, even with `--mlir-print-op-generic` Are you saying that there are options for `--mlir-print-region-generic` etc that always print the block name?

Reply:
> The default generic format always prints the block name.
This is not true, see this example from MLIR:
Initial program: 
```
func.func @std_if_else(%arg0: i1) {
  scf.if %arg0 {} else {}
  return
}
```
Parsed with `mlir-opt example.mlir`:
```
  func.func @std_if_else(%arg0: i1) {
    scf.if %arg0 {
    } else {
    }
    return
  }
```
Parsed with `mlir-opt example.mlir --mlir-print-op-generic`:
```
  "func.func"() ({
  ^bb0(%arg0: i1):
    "scf.if"(%arg0) ({
      "scf.yield"() : () -> ()
    }, {
      "scf.yield"() : () -> ()
    }) : (i1) -> ()
    "func.return"() : () -> ()
  }) {function_type = (i1) -> (), sym_name = "std_if_else"} : () -> ()
```

This is special cased for single block regions. For regions with multiple blocks the name is always printed in both formats.

Reply:
Wait, then there is something weird going on.
On main, I have
```
module {}
```
being printed as:
```
"builtin.module"() ({
^bb0:
}) : () -> ()
```

Reply:
The difference here is `scf.if` is defined with `SingleBlockImplicitTerminator<"scf::YieldOp">`, which triggers this printing.
see [here](https://github.com/opencompl/llvm-project/blob/abb32a95f75876b8ec5250a0e2ce541073cc3cc2/mlir/include/mlir/Dialect/SCF/IR/SCFOps.td#L574)

module on the other hand is not, see [here](https://github.com/opencompl/llvm-project/blob/abb32a95f75876b8ec5250a0e2ce541073cc3cc2/mlir/include/mlir/IR/BuiltinOps.td#L33)

This should really really not influence the generic printing in MLIR, but sadly it does.

Reply:
hmm, potentially an issue for later